include/aws/common/string.h

```diff
@@ -58,31 +58,72 @@ struct aws_string {
 #endif
 
 /**
- * Equivalent to str->bytes. Here for legacy reaasons.
+ * Evaluates the set of properties that define the shape of all valid aws_string structures.
+ * It is also a cheap check, in the sense it run in constant time (i.e., no loops or recursion).
+ */
+AWS_STATIC_IMPL bool aws_string_is_valid(const struct aws_string *str) {
+    return str && AWS_MEM_IS_READABLE(&str->bytes[0], str->len + 1) && str->bytes[str->len] == 0;
+}
+
+/**
+ * Best-effort checks aws_string invariants, when the str->len is unknown
+ */
+AWS_STATIC_IMPL bool aws_c_string_is_valid(const struct aws_string *str) {
+    /* Knowing the actual length to check would require strlen(), which is
+     * a) linear time in the length of the string
+     * b) could already cause a memory violation for a non-zero-terminated string.
+     * But we know that a c-string must have at least one character, to store the null terminator
+     */
+    return str && AWS_MEM_IS_READABLE(str, 1);
+}
+
+/**
+ * Equivalent to str->bytes. Here for legacy reasons.
  */
 AWS_STATIC_IMPL const uint8_t *aws_string_bytes(const struct aws_string *str) {
+    AWS_FATAL_ASSERT(aws_string_is_valid(str));
     return str->bytes;
 }
 
 /**
  * Returns true if bytes of string are the same, false otherwise.
  */
 AWS_STATIC_IMPL bool aws_string_eq(const struct aws_string *a, const struct aws_string *b) {
-    AWS_FATAL_ASSERT(a != NULL && b != NULL);
+    AWS_PRECONDITION(aws_string_is_valid(a));
+    AWS_PRECONDITION(aws_string_is_valid(b));
+    if (a == b) {
+        return true;
+    }
+    if (a == NULL || b == NULL) {
+        return false;
+    }
     return aws_array_eq(a->bytes, a->len, b->bytes, b->len);
 }
 
 /**
  * Returns true if bytes of string are equivalent, using a case-insensitive comparison.
  */
 AWS_STATIC_IMPL bool aws_string_eq_ignore_case(const struct aws_string *a, const struct aws_string *b) {
+    AWS_PRECONDITION(aws_string_is_valid(a));
+    AWS_PRECONDITION(aws_string_is_valid(b));
+    if (a == b) {
+        return true;
+    }
+    if (a == NULL || b == NULL) {
+        return false;
+    }
     return aws_array_eq_ignore_case(a->bytes, a->len, b->bytes, b->len);
 }
 
 /**
  * Returns true if bytes of string and cursor are the same, false otherwise.
  */
 AWS_STATIC_IMPL bool aws_string_eq_byte_cursor(const struct aws_string *str, const struct aws_byte_cursor *cur) {
+    AWS_PRECONDITION(aws_string_is_valid(str));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(cur));
```

Comment:
This precondition is too strict, as both `str` and `cur` are allowed to be NULL (based on the branch below).